File: src/modules/userConfigManager.js
const fs = require('fs');
const path = require('path');
const moment = require('moment');

class UserConfigManager {
  constructor() {
    this.storePath = path.join(__dirname, '..', 'store', 'userConfigs.json');
    this.userConfigs = {};
    this.inactivityDays = 7;
    this.load();
  }

  load() {
    try {
      if (fs.existsSync(this.storePath)) {
        const data = fs.readFileSync(this.storePath, 'utf8');
        this.userConfigs = JSON.parse(data);
      }
    } catch (error) {
      this.userConfigs = {};
    }
  }

  save() {
    try {
      const dir = path.dirname(this.storePath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.writeFileSync(this.storePath, JSON.stringify(this.userConfigs, null, 2));
    } catch (error) {
      console.error('Failed to save user configs:', error.message);
    }
  }

  createUserConfig(userId, username) {
    this.userConfigs[userId] = {
      username: username,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      lastActivity: new Date().toISOString(),
      github: {
        pat: null,
        repoPath: null,
        repoUrl: null,
        branch: 'main',
        remote: 'origin'
      },
      discord: {
        dmEnabled: true
      },
      preferences: {
        defaultCommitMessage: '',
        timezone: 'UTC'
      }
    };
    this.save();
    return this.userConfigs[userId];
  }

  getUserConfig(userId) {
    return this.userConfigs[userId] || null;
  }

  updateUserConfig(userId, updates) {
    if (!this.userConfigs[userId]) {
      return null;
    }
    
    this.userConfigs[userId] = {
      ...this.userConfigs[userId],
      ...updates,
      updatedAt: new Date().toISOString()
    };
    this.save();
    return this.userConfigs[userId];
  }

  updateLastActivity(userId) {
    if (this.userConfigs[userId]) {
      this.userConfigs[userId].lastActivity = new Date().toISOString();
      this.save();
    }
  }

  cleanupInactiveUsers(inactivityDays = null) {
    const days = inactivityDays || this.inactivityDays;
    const cutoff = moment().subtract(days, 'days');
    let cleaned = 0;

    for (const [userId, config] of Object.entries(this.userConfigs)) {
      if (config.lastActivity) {
        const lastActive = moment(config.lastActivity);
        if (lastActive.isBefore(cutoff)) {
          delete this.userConfigs[userId];
          cleaned++;
        }
      }
    }

    if (cleaned > 0) {
      this.save();
      console.log(`Cleaned up ${cleaned} inactive users`);
    }

    return cleaned;
  }

  setGitHubPAT(userId, pat) {
    if (!this.userConfigs[userId]) {
      return { success: false, error: 'User config not found' };
    }
    this.userConfigs[userId].github.pat = pat;
    this.userConfigs[userId].updatedAt = new Date().toISOString();
    this.save();
    return { success: true };
  }

  getGitHubPAT(userId) {
    return this.userConfigs[userId]?.github?.pat || null;
  }

  setRepoConfig(userId, repoPath, branch = 'main', remote = 'origin') {
    if (!this.userConfigs[userId]) {
      return { success: false, error: 'User config not found' };
    }
    this.userConfigs[userId].github.repoPath = repoPath;
    this.userConfigs[userId].github.branch = branch;
    this.userConfigs[userId].github.remote = remote;
    this.userConfigs[userId].updatedAt = new Date().toISOString();
    this.save();
    return { success: true };
  }

  getRepoConfig(userId) {
    return this.userConfigs[userId]?.github || null;
  }

  hasGitHubSetup(userId) {
    const config = this.userConfigs[userId]?.github;
    return !!(config?.pat && config?.repoPath);
  }

  deleteUserConfig(userId) {
    if (this.userConfigs[userId]) {
      delete this.userConfigs[userId];
      this.save();
      return true;
    }
    return false;
  }

  listUsers() {
    return Object.keys(this.userConfigs).map(userId => ({
      userId,
      username: this.userConfigs[userId].username,
      hasGitHub: this.hasGitHubSetup(userId),
      updatedAt: this.userConfigs[userId].updatedAt
    }));
  }
}

module.exports = new UserConfigManager();

File: BOT/python-bot/modules/user_config_manager.py
import os
import json
from datetime import datetime

class UserConfigManager:
    def __init__(self):
        self.store_path = os.path.join('store', 'user_configs.json')
        self.user_configs = {}
        self.session_inactivity_days = int(os.getenv('SESSION_INACTIVITY_DAYS', '7'))

    def initialize(self):
        os.makedirs('store', exist_ok=True)
        self.load()

    def load(self):
        try:
            if os.path.exists(self.store_path):
                with open(self.store_path, 'r') as f:
                    self.user_configs = json.load(f)
        except Exception:
            self.user_configs = {}

    def save(self):
        try:
            os.makedirs(os.path.dirname(self.store_path), exist_ok=True)
            with open(self.store_path, 'w') as f:
                json.dump(self.user_configs, f, indent=2)
        except Exception as e:
            print(f'Failed to save user configs: {e}')

    def create_user_config(self, user_id, username):
        self.user_configs[user_id] = {
            'username': username,
            'created_at': datetime.now().isoformat(),
            'updated_at': datetime.now().isoformat(),
            'last_activity': datetime.now().isoformat(),
            'github': {
                'pat': None,
                'repo_path': None,
                'repo_url': None,
                'branch': 'main',
                'remote': 'origin'
            },
            'discord': {
                'dm_enabled': True
            },
            'preferences': {
                'default_commit_message': '',
                'timezone': 'UTC'
            }
        }
        self.save()
        return self.user_configs[user_id]

    def get_user_config(self, user_id):
        return self.user_configs.get(user_id)

    def update_user_config(self, user_id, updates):
        if user_id not in self.user_configs:
            return None
        
        self.user_configs[user_id] = {
            **self.user_configs[user_id],
            **updates,
            'updated_at': datetime.now().isoformat()
        }
        self.save()
        return self.user_configs[user_id]

    def update_last_activity(self, user_id):
        if user_id in self.user_configs:
            self.user_configs[user_id]['last_activity'] = datetime.now().isoformat()
            self.save()

    def set_github_pat(self, user_id, pat):
        if user_id not in self.user_configs:
            return {'success': False, 'error': 'User config not found'}
        self.user_configs[user_id]['github']['pat'] = pat
        self.user_configs[user_id]['updated_at'] = datetime.now().isoformat()
        self.save()
        return {'success': True}

    def get_github_pat(self, user_id):
        return self.user_configs.get(user_id, {}).get('github', {}).get('pat')

    def set_repo_config(self, user_id, repo_path, branch='main', remote='origin'):
        if user_id not in self.user_configs:
            return {'success': False, 'error': 'User config not found'}
        self.user_configs[user_id]['github']['repo_path'] = repo_path
        self.user_configs[user_id]['github']['branch'] = branch
        self.user_configs[user_id]['github']['remote'] = remote
        self.user_configs[user_id]['updated_at'] = datetime.now().isoformat()
        self.save()
        return {'success': True}

    def get_repo_config(self, user_id):
        return self.user_configs.get(user_id, {}).get('github')

    def has_github_setup(self, user_id):
        config = self.user_configs.get(user_id, {}).get('github')
        return bool(config and config.get('pat') and config.get('repo_path'))

    def get_inactive_users(self, days=None):
        if days is None:
            days = self.session_inactivity_days
        
        cutoff = datetime.now().timestamp() - (days * 24 * 60 * 60)
        inactive = []
        
        for user_id, config in self.user_configs.items():
            last_activity = config.get('last_activity')
            if last_activity:
                try:
                    activity_time = datetime.fromisoformat(last_activity).timestamp()
                    if activity_time < cutoff:
                        inactive.append(user_id)
                except ValueError:
                    pass
        
        return inactive

    def delete_user_config(self, user_id):
        if user_id in self.user_configs:
            del self.user_configs[user_id]
            self.save()
            return True
        return False

    def list_users(self):
        return [
            {
                'user_id': user_id,
                'username': config.get('username'),
                'has_github': self.has_github_setup(user_id),
                'updated_at': config.get('updated_at'),
                'last_activity': config.get('last_activity')
            }
            for user_id, config in self.user_configs.items()
        ]

    def get_all_scheduled_jobs(self):
        from modules.scheduler_service import SchedulerService
        return SchedulerService.list_all_jobs()
